### DISCUSSION TITLE
How to run example under windows xp

### DISCUSSION THREAD

**Discussion by xixioba** [ocornut/imgui#8039] [New Users Build/Link/Run issues ONLY!]
I compiled the sample code under [msys2 mingw32](https://www.msys2.org/).
```
g++ -DUNICODE -I../.. -I../../backends main.cpp ../../backends/imgui_impl_opengl3.cpp ../../backends/imgui_impl_win32.cpp ../../imgui*.cpp -o Debug/example_win32_opengl3.exe --static -mwindows -lopengl32 -lgdi32 -ldwmapi -lopengl32 -lgdi32 -ldwmapi
```
When I first ran it, I got the error `dwmapi.dll not found`. I downloaded this dll to fix this problem.

Then I got the error `point_ftol2 could not be located int the dynamic link library msvcrt.dll.`
![截屏2024-10-05 15 29 12](https://github.com/user-attachments/assets/4a700f24-e427-439e-9dd7-3a8bef11b717)

 I don't know how to fix it.

**Comment by nicolasnoble** [comment]
Windows XP has long been deprecated by Microsoft, and msys2 / mingw dropped support for it some time ago. It's not expected to work there, and you're not supposed to use this ancient operating system anymore.

**Comment by xixioba** [comment]
@nicolasnoble For some reason, I wanted to test the platform compatibility of imgui and egui.

After some attempts, the good news is that [**example_glfw_opengl2**](https://github.com/ocornut/imgui/tree/master/examples/example_glfw_opengl2) can be successfully run in **VMware** with [**Windows XP SP3**](https://archive.org/details/WinXPProSP3x86) 

Anyone who searches here can refer to my solution(**both Mingw32 and Msvc** are available)

- [**MSYS2 Mingw32**](https://www.msys2.org/), install **gcc**, **make** and **glfw** dependencies ([link](https://packages.msys2.org/queue) for more packages)

  ```
  pacman -S make mingw-w64-i686-gcc mingw-w64-i686-glfw
  ```
  Modify the **makefile** and `make`, thats all
  ```
  CXXFLAGS = -std=c++11 -I$(IMGUI_DIR) -I$(IMGUI_DIR)/backends
  CXXFLAGS += -g -Wall -Wformat -D _USING_V110_SDK71_ -Wl,--subsystem,windows -Wl,--subsystem,console -static -static-libgcc -static-libstdc++
  LIBS =
  ```
- [**Msvc**](https://learn.microsoft.com/en-us/visualstudio/releases/2019/history#release-dates-and-build-numbers)(Visual Studio 2019 16.9.26 is the last version that supports XP)
  If you only use the command line, just download and install [**BuildTools**](https://download.visualstudio.microsoft.com/download/pr/245e99d9-73d8-4db6-84eb-493b0c059e15/b2fd18b4c66d507d50aced118be08937da399cd6edb3dc4bdadf5edc139496d4/vs_BuildTools.exe) and **check Windows XP Support**
  Modify ths **build_win32.bat** and `build_win32.bat`, thats all
  ```
  cl /nologo /Zi /MD /utf-8 /D_USING_V110_SDK71_ %INCLUDES% %SOURCES% /Fe%OUT_DIR%/%OUT_EXE%.exe /Fo%OUT_DIR%/ /link /SUBSYSTEM:WINDOWS,5.01 /SUBSYSTEM:CONSOLE,5.01 %LIBS%
  ```
  
  Now you can copy the generated executable program to Windows XP, run it, you can see a simple demonstration, have fun!
  
![截屏2024-10-07 14 43 00](https://github.com/user-attachments/assets/7dcd47a4-ec48-4e4b-a76d-456ffc8dc382)

**Reply by LTVA1** [reply]
Hello, msvc doesn't work, on virtual machine with 32-bit win xp professional service pack 3 it doesn't launch, writes an error about releasesrwlockexclusive not available in kernel32.dll file

**Reply by LTVA1** [reply]
For mingw32 it gives this lol
<img width="1110" height="273" alt="image" src="https://github.com/user-attachments/assets/b44bedcb-18c4-4096-b0d6-77ca3a6eac60" />

**Reply by LTVA1** [reply]
Okay, it works finally! You need to use mingw32.exe to compile it!

**Reply by LTVA1** [reply]
Although it only works for 32-bit windows xp, probably because 64 bit one doesn't have service pack 3. How to compile for windows xp 64-bit? I don't see software renderer example there, but it's probably the only way to do it. I have seen imgui app compiled for Windows 2000 though

**Reply by xixioba** [reply]
I don't have Windows XP x64 Service Pack 3 so I haven't tried it yet. If you have any good results, please post them. It will be interesting.

**Reply by LTVA1** [reply]
Quick googling told me there never was any service pack 3 for 64-bit windows xp versions. I probably need to try software renderer for those old systems but I have very little experience with those build systems and old Windows support. Maybe you have some example? I know a dear imgui app which states it can be run even on Windows 2000, but build system is based on cmake and some Linux scripts I can't run on Windows even with mingw... And app is big so I can't easily boil it down into an example

**Reply by LTVA1** [reply]
Maybe you will be able to condense it into nice one-file example? App is https://github.com/tildearrow/furnace

**Reply by xixioba** [reply]
Looks cool, but this is just my hobby outside of work, so it may not be of any help to you. I will let you know if there is more progress.
